contracts/FreeMint.sol

```diff
@@ -0,0 +1,184 @@
+        require(
+            recipients.length == amounts.length,
+            "Recipients and amounts length mismatch"
+        );
+
+        uint256 totalAmount = 0;
+        for (uint256 i = 0; i < amounts.length; i++) {
+            require(amounts[i] > 0, "Invalid amount");
+            totalAmount += amounts[i];
+        }
+
+        require(
+            _nextTokenId + totalAmount <= maxSupply,
+            "Max supply reached"
+        );
+
+        for (uint256 i = 0; i < recipients.length; i++) {
+            // Allow unlimited minting if mintLimit is 0
+            if (mintLimit != 0) {
+                require(
+                    balanceOf(recipients[i]) + amounts[i] <= mintLimit,
+                    "Exceeds mint limit"
+                );
+            }
+            for (uint256 j = 0; j < amounts[i]; j++) {
+                _safeMint(recipients[i], _nextTokenId++);
+            }
+        }
+    }
+
+    // ---- The following functions are overrides required by Solidity. ---- //
+
+    function _update(
+        address to,
+        uint256 tokenId,
+        address auth
+    ) internal override(ERC721, ERC721Pausable) returns (address) {
+        return super._update(to, tokenId, auth);
+    }
+
+    // ---------- Setter Functions ---------- //
+
+    //@notice to contractUri
+    function setContractURI(
+        string memory _newBaseContractURI
+    ) public onlyOwner {
+        baseContractURI = _newBaseContractURI;
+    }
+
+    //@notice to tokenUri
+    function setBaseURI(string memory _newBaseURI) public onlyOwner {
+        baseURI = _newBaseURI;
+    }
+
+    //@notice to tokenUri
+    function setBaseExtension(
+        string memory _newBaseExtension
+    ) public onlyOwner {
+        baseExtension = _newBaseExtension;
+    }
+
+    //@notice maxSupply
+    function setMaxSupply(uint256 _newMaxSupply) public onlyOwner {
+        maxSupply = _newMaxSupply;
+    }
+
+    //@notice mintLimit
+    function setMintLimit(uint256 _newMintLimit) public onlyOwner {
+        mintLimit = _newMintLimit;
+    }
+
+    //@notice totalSupply
+    function totalSupply() public view returns (uint256) {
+        return _nextTokenId;
+    }
+
+    // ---------- Getter Functions ---------- //
+
+
+    //@notice tokenURI override
+    function tokenURI(
+        uint256 tokenId
+    ) public view override(ERC721, ERC721URIStorage) returns (string memory) {
+        _requireOwned(tokenId);
+        require(tokenId < _nextTokenId, "Token ID is invalid");
+
+        // Check if the last character in baseURI is a slash.
+        // if there is no slash, use baseURI without token ID
+        if (bytes(baseURI)[bytes(baseURI).length - 1] != bytes("/")[0]) {
+            return baseURI;
+        }
+
+        return
+            bytes(baseURI).length > 0
+                ? string(
+                    abi.encodePacked(
+                        baseURI,
+                        (tokenId % 5 == 0 ? 5 : tokenId % 5).toString(),
```

Comment:
Why 5?

Reply:
good catch... In this contract it should be pure `tokenId`.  No numbers

Reply:
fixed and added test for this case